File: src/python/parc.py
"""
PARC Reference Engine (Python)
UTF-PARC 1.0 Compatible
Author: Your Name
License: MIT

This file provides:
- parc_vector(text): produce a UTF-PARC cognitive state vector
- clamp(x): ensure values fall in [0, 1]
- normalize(c, m, f, k): enforce UTF-PARC constraints
- parc_update(c, m, f, k): optional iterative update (PARC dynamics)

This implementation intentionally uses a simple heuristic model so that it
runs anywhere with zero dependencies. You can replace encode_text() with
an embedding model, classifier, or domain-specific logic.
"""

import math
import random


# -----------------------------
# Utility Functions
# -----------------------------
def clamp(x: float) -> float:
    """Clamp a float to [0, 1]."""
    return max(0.0, min(1.0, float(x)))


# -----------------------------
# Heuristic Encoding Layer
# -----------------------------
def encode_text(text: str):
    """
    Convert text into initial (c0, m0, f0, k0) seeds.
    This is the only part you'd swap out for a real model.

    Current heuristic:
    - correctness increases with definitional markers ("is", "means", "because")
    - misconception increases with negation or contradiction markers
    - confidence is partly random but anchored
    - fog is derived during normalization (UTF-PARC rule)
    """

    t = text.lower().strip()

    # word lists (simple placeholders)
    correct_words = ["is", "are", "means", "because"]
    mis_words = ["not", "never", "wrong", "unless"]

    tokens = t.split()
    length = max(1, len(tokens))

    # Very rough signals
    c0 = sum(1 for w in tokens if w in correct_words) / length
    m0 = sum(1 for w in tokens if w in mis_words) / length

    # raw fog (will be replaced later by UTF-PARC rule)
    f0 = 1 - max(c0, m0)

    # semi-randomized confidence with meaningful drift
    k0 = 0.3 + 0.7 * random.random()

    return float(c0), float(m0), float(f0), float(k0)


# -----------------------------
# UTF-PARC Normalization
# -----------------------------
def normalize(c0, m0, f0, k0):
    """
    Enforce UTF-PARC 1.0 constraints:
      - clamp all values
      - normalize c + m ≤ 1
      - recompute f = 1 - max(c, m)
      - return (c, m, f, k)
    """

    # clamp first
    c = clamp(c0)
    m = clamp(m0)
    k = clamp(k0)

    # enforce c + m ≤ 1
    total = c + m
    if total > 1.0:
        c = c / total
        m = m / total

    # UTF-PARC fog rule
    f = 1.0 - max(c, m)

    return c, m, f, k


# -----------------------------
# Optional: PARC Temporal Update
# -----------------------------
def parc_update(c, m, f, k, gamma=0.3, delta=0.1, beta=0.5, rho=0.5, steps=3):
    """
    Apply PARC dynamical update equations.
    Used when tracking learner reasoning over time.
    """

    for _ in range(steps):

        gamma_eff = gamma * (1 - m)
        delta_eff = delta + beta * c

        # update correctness + misconception
        c_new = c + gamma_eff * (1 - c - m)
        m_new = m * (1 - delta_eff)

        # renormalize if needed
        if c_new + m_new > 1:
            total = c_new + m_new
            c_new /= total
            m_new /= total
            f_new = 0.0
        else:
            f_new = 1.0 - max(c_new, m_new)

        # confidence follows c - m
        k_new = k + rho * (c_new - m_new - k)

        c, m, f, k = c_new, m_new, f_new, k_new

    return c, m, f, k


# -----------------------------
# Public API
# -----------------------------
def parc_vector(text: str, apply_updates=True):
    """
    Produce the final UTF-PARC 1.0 vector for a given text.
    Steps:
      1. Encode via heuristic (or custom model)
      2. Normalize per UTF-PARC spec
      3. Optionally apply PARC update equations
    """

    c0, m0, f0, k0 = encode_text(text)
    c, m, f, k = normalize(c0, m0, f0, k0)

    if apply_updates:
        c, m, f, k = parc_update(c, m, f, k)

    return {
        "c": round(float(c), 4),
        "m": round(float(m), 4),
        "f": round(float(f), 4),
        "k": round(float(k), 4),
    }


# -----------------------------
# Command Line Usage
# -----------------------------
if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1:
        text = " ".join(sys.argv[1:])
        print(parc_vector(text))
    else:
        print("Usage: python parc.py \"your text here\"")

File: src/js/parc.mjs
// src/js/parc.mjs
// PARC Reference Engine (JavaScript / ES Module)
// UTF-PARC 1.0 Compatible
// License: MIT

// -----------------------------
// Utility
// -----------------------------

/**
 * Clamp a number into [0, 1].
 * @param {number} x
 * @returns {number}
 */
export function clamp(x) {
  const n = Number(x);
  if (Number.isNaN(n)) return 0;
  return Math.max(0, Math.min(1, n));
}

/**
 * Simple seeded-ish random wrapper if you ever want determinism.
 * For now, we just expose Math.random().
 */
function rand() {
  return Math.random();
}

// -----------------------------
// Heuristic Encoding Layer
// -----------------------------

/**
 * Convert text into initial (c0, m0, f0, k0) seeds.
 * This mirrors the Python encode_text() function.
 *
 * Heuristic:
 *  - correctness (c0) bumps on definitional markers: "is", "are", "means", "because"
 *  - misconception (m0) bumps on negation-like markers: "not", "never", "wrong", "unless"
 *  - fog (f0) is 1 - max(c0, m0) (will be recomputed later anyway)
 *  - confidence (k0) is 0.3–1.0 with a random drift
 *
 * @param {string} text
 * @returns {[number, number, number, number]} tuple [c0, m0, f0, k0]
 */
export function encodeText(text) {
  const t = String(text).toLowerCase().trim();
  const tokens = t.split(/\s+/).filter(Boolean);
  const len = Math.max(1, tokens.length);

  const correctWords = ["is", "are", "means", "because"];
  const misWords = ["not", "never", "wrong", "unless"];

  let c0 = 0;
  let m0 = 0;

  for (const w of tokens) {
    if (correctWords.includes(w)) c0 += 1;
    if (misWords.includes(w)) m0 += 1;
  }

  c0 = c0 / len;
  m0 = m0 / len;

  const f0 = 1 - Math.max(c0, m0);
  const k0 = 0.3 + 0.7 * rand();

  return [c0, m0, f0, k0];
}

// -----------------------------
// UTF-PARC Normalization
// -----------------------------

/**
 * Enforce UTF-PARC 1.0 rules:
 *  - clamp all values to [0,1]
 *  - ensure c + m ≤ 1
 *  - recompute fog as f = 1 - max(c, m)
 *
 * @param {number} c0
 * @param {number} m0
 * @param {number} f0   (ignored except for debugging)
 * @param {number} k0
 * @returns {[number, number, number, number]} [c, m, f, k]
 */
export function normalize(c0, m0, f0, k0) {
  let c = clamp(c0);
  let m = clamp(m0);
  let k = clamp(k0);

  const total = c + m;
  if (total > 1) {
    c = c / total;
    m = m / total;
  }

  const f = 1 - Math.max(c, m);

  return [c, m, f, k];
}

// -----------------------------
// PARC Temporal Update
// -----------------------------

/**
 * Apply PARC dynamical update equations (optional).
 * Used when tracking reasoning over time.
 *
 * Mirrors Python parc_update().
 *
 * @param {number} c
 * @param {number} m
 * @param {number} f
 * @param {number} k
 * @param {object}  [opts]
 * @param {number}  [opts.gamma=0.3]
 * @param {number}  [opts.delta=0.1]
 * @param {number}  [opts.beta=0.5]
 * @param {number}  [opts.rho=0.5]
 * @param {number}  [opts.steps=3]
 * @returns {[number, number, number, number]} [c, m, f, k]
 */
export function parcUpdate(
  c,
  m,
  f,
  k,
  opts = {}
) {
  let {
    gamma = 0.3,
    delta = 0.1,
    beta = 0.5,
    rho = 0.5,
    steps = 3,
  } = opts;

  for (let i = 0; i < steps; i++) {
    const gammaEff = gamma * (1 - m);
    const deltaEff = delta + beta * c;

    let cNew = c + gammaEff * (1 - c - m);
    let mNew = m * (1 - deltaEff);

    let fNew;
    if (cNew + mNew > 1) {
      const total = cNew + mNew;
      cNew = cNew / total;
      mNew = mNew / total;
      fNew = 0;
    } else {
      fNew = 1 - Math.max(cNew, mNew);
    }

    const kNew = k + rho * (cNew - mNew - k);

    c = cNew;
    m = mNew;
    f = fNew;
    k = kNew;
  }

  return [c, m, f, k];
}

// -----------------------------
// Public API
// -----------------------------

/**
 * Produce the final UTF-PARC 1.0 vector for a given text.
 *
 * @param {string} text
 * @param {object} [opts]
 * @param {boolean} [opts.applyUpdates=true]
 * @returns {{c:number, m:number, f:number, k:number}}
 */
export function parcVector(text, opts = {}) {
  const { applyUpdates = true } = opts;

  const [c0, m0, f0, k0] = encodeText(text);
  let [c, m, f, k] = normalize(c0, m0, f0, k0);

  if (applyUpdates) {
    [c, m, f, k] = parcUpdate(c, m, f, k);
  }

  return {
    c: Number(c.toFixed(4)),
    m: Number(m.toFixed(4)),
    f: Number(f.toFixed(4)),
    k: Number(k.toFixed(4)),
  };
}

// Default export for convenience
export default {
  clamp,
  encodeText,
  normalize,
  parcUpdate,
  parcVector,
};
